gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -504,6 +506,189 @@ def _get_variant_calls(self, record, formats):
+  class - we could only use a single pipe, but it will divert the parsers.
+  """
+
+  def __init__(self,
+               file_name,  # type: str
+               range_tracker,  # type: range_trackers.OffsetRangeTracker
+               compression_type,  # type: str
+               allow_malformed_records,  # type: bool
+               file_pattern=None,  # type: str
+               representative_header_lines=None,  # type:  List[str]
+               splittable_bgzf=False,  # type: bool
+               **kwargs  # type: **str
+              ):
+    # type: (...) -> None
+    super(PySamParser, self).__init__(file_name,
+                                      range_tracker,
+                                      file_pattern,
+                                      compression_type,
+                                      allow_malformed_records,
+                                      representative_header_lines,
+                                      **kwargs)
+    # This member will be properly initiated in _init_with_header().
+    self._vcf_reader = None
+    self._to_child = None
+    self._original_info_list = None
+
+  def _init_child_thread(self, p_read, c_write, c_read, p_write, header_lines):
+    # Child thread's task is to populate data into the pipe that feeds
+    # VariantFile class - first by populating all of the header lines, and then
+    # by redirecting the data received from the second pipe.
+
+    os.close(p_read)
+    os.close(p_write)
+    from_parent = os.fdopen(c_read)
+    to_pysam = os.fdopen(c_write, 'w')
+    # Feed the data with a break line, since we read data line by line - if
+    # we would use the conventional ``read()`` method, both pipes would lock,
+    # while waiting for an EOF signal.
+    for header in header_lines:
+      to_pysam.write(header.strip() + '\n')
+    to_pysam.flush()
+
+    while True:
+      text_line = from_parent.readline()
+      if not text_line:
+        break
+      to_pysam.write(text_line)
+      to_pysam.flush()
+
+    from_parent.close()
+    to_pysam.close()
+    sys.exit(0)
+
+  def _init_with_header(self, header_lines):
+    # PySam requires a version header to be present, so add one if absent.
+    if header_lines and not header_lines[0].startswith(
+        FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='')):
+      header_lines.insert(0, FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='4.0'))
+
+    # First pipe is responsible for supplying data from child thread into the
+    # VariantFile, since it only takes an actual file descriptor as its input.
+    p_read, c_write = os.pipe()
+    # Since vcf_parser processes 1 line at a time, a second pipe is used to send
+    # 1 record at a time to the child thread, to pass it downstream back to
+    # parent thread via the first pipe/VariantFile class.
+    c_read, p_write = os.pipe()
+
+    processid = os.fork()
+    if processid:
+      # Parent Thread
+      os.close(c_read)
```

Comment:
Why bother closing these?

Reply:
It makes it less clear in my mind: you are touching things that you don't need to touch